compiler/rustc_hir_analysis/src/check/region.rs

```diff
@@ -887,3 +897,108 @@ pub fn region_scope_tree(tcx: TyCtxt<'_>, def_id: DefId) -> &ScopeTree {
 
     tcx.arena.alloc(scope_tree)
 }
+
+/// Check for shared reference or mutable reference of static
+pub fn maybe_static_ref(
+    tcx: TyCtxt<'_>,
+    block: Option<Block<'_>>,
+    f: Option<(FnKind<'_>, hir::BodyId)>,
+) {
+    fn check_stmts(tcx: TyCtxt<'_>, hir_id: HirId, stmts: &[Stmt<'_>], safe: bool) {
+        for stmt in stmts {
+            if let hir::StmtKind::Local(loc) = stmt.kind
+                && let Some(init) = loc.init
+                && let hir::ExprKind::AddrOf(borrow_kind, _, expr) = init.kind
+                && matches!(borrow_kind, BorrowKind::Ref)
+                && let hir::ExprKind::Path(qpath) = expr.kind
+                && let hir::QPath::Resolved(_, path) = qpath
+                && let hir::def::Res::Def(def_kind, _) = path.res
+                && let hir::def::DefKind::Static(mt) = def_kind
+            {
+                handle_static_ref(
+                    tcx,
+                    init.span,
+                    qpath_to_string(&qpath),
+                    loc.span.edition().at_least_rust_2024(),
+                    matches!(mt, Mutability::Mut),
+                    safe,
+                    hir_id,
+                );
+            }
+        }
+    }
+
+    fn handle_static_ref(
+        tcx: TyCtxt<'_>,
+        span: Span,
+        var: String,
+        e2024: bool,
+        mutable: bool,
+        safe: bool,
+        hir_id: HirId,
+    ) {
+        if e2024 && mutable && !safe {
+            tcx.sess.emit_err(errors::StaticMutRef { span, var });
+        } else if !e2024 {
+            let (mut shared, mut mutable_str) = ("", "");
+            let (label, sugg) = if mutable {
+                mutable_str = "mutable ";
+                (errors::RefOfStaticLabel::Mut { span }, errors::RefOfStaticSugg::Mut { span, var })
+            } else {
+                shared = "shared ";
+                (
+                    errors::RefOfStaticLabel::Shared { span },
+                    errors::RefOfStaticSugg::Shared { span, var },
+                )
+            };
+            if mutable && !safe {
```

Comment:
this if has the same content for both the if and the else if body. Maybe it would be better to collapse it into one `if (mutable && !safe) || !mutable`, or alternatively `if matches!((true, false) | (false, _), (mutable, safe))`.